device/Microsoft.Azure.Devices.Client/Transport/Mqtt/MqttTransportHandler.cs

```diff
@@ -454,26 +454,34 @@ public void OnConnected()
 
         public async void OnMessageReceived(Message message)
         {
-            if ((this.State & TransportState.Open) == TransportState.Open)
+            // Added Try-Catch to avoid unknown thread exception
+            // after running for more than 24 hours
+            try
             {
-                if (message.MqttTopicName.StartsWith(twinResponseTopicPrefix))
-                {
-                    twinResponseEvent(message);
-                }
-                else if (message.MqttTopicName.StartsWith(twinPatchTopicPrefix))
-                {
-                    await HandleIncomingTwinPatch(message).ConfigureAwait(false);
-                }
-                else if (message.MqttTopicName.StartsWith(methodPostTopicPrefix))
+                if ((this.State & TransportState.Open) == TransportState.Open)
                 {
-                    await HandleIncomingMethodPost(message).ConfigureAwait(false);
-                }
-                else
-                {
-                    this.messageQueue.Enqueue(message);
-                    this.receivingSemaphore.Release();
+                    if (message.MqttTopicName.StartsWith(twinResponseTopicPrefix))
+                    {
+                        twinResponseEvent(message);
+                    }
+                    else if (message.MqttTopicName.StartsWith(twinPatchTopicPrefix))
+                    {
+                        await HandleIncomingTwinPatch(message).ConfigureAwait(false);
+                    }
+                    else if (message.MqttTopicName.StartsWith(methodPostTopicPrefix))
+                    {
+                        await HandleIncomingMethodPost(message).ConfigureAwait(false);
+                    }
+                    else
+                    {
+                        this.messageQueue.Enqueue(message);
+                        this.receivingSemaphore.Release();
+                    }
                 }
             }
+            catch(Exception)
```

Comment:
Catching the exception is the right fix but after this is caught, we need to flow the exception to the application, not hide it.
This would be similar to https://github.com/Azure/azure-iot-sdk-csharp/blob/master/provisioning/transport/mqtt/src/ProvisioningChannelHandlerAdapter.cs#L305
 
I believe that in this context the code needs to call `this.OnError(ex)` to move the state machine into a faulted state and cleanup.

Reply:
Also we should catch non-fatal exceptions:
`            catch (Exception ex) when (!ex.IsFatal())`